### PR TITLE
add D generated files to order-only deps

### LINKED ISSUE #10463: Missing implicit dependency to generated code
**Describe the bug**
When using the result of `custom_target` as source to a build target, an implicit dependency (or order-only dependency?) should be added to each object file to ensure that the generated code is built before the other sources, because the other sources may depend on the generated code.

**To Reproduce**
Try to build the project https://github.com/rtbo/meson-dep-test (meson.build copied hereafter).
Compilation of `lib.d` races with compilation of `generated.d` and cannot find module `generated`.
If I manually add `generated.o` as implicit dependency of `lib.o` in `ninja.build`, everything is file.

```meson
project('meson-dep-test', 'd')

generator = executable('generator', 'generator.d')

generated = custom_target('generated',
    capture: true,
    output: 'generated.d',
    input: 'input.txt',
    command: [
        generator, '@INPUT@'
    ]
)

lib = static_library('lib', generated, 'lib.d',
    include_directories: include_directories('.'),
)
```

**Expected behavior**
Reliable build without race

**system parameters**
Arch Linux
meson master branch (0.62.99)
Python 3.10.4
Ninja 1.11.0

### PR DESCRIPTION
**Pull request by rtbo**
fixes #10463

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -845,6 +850,8 @@ def generate_target(self, target):
             compiled_sources.append(s)
             source2object[s] = o
             obj_list.append(o)
+            if s.endswith('.d'):
```
Comment (rtbo) [inline mesonbuild/backend/ninjabackend.py:853]: Is there a predefined routine to identify the language of a source file?

Comment (eli-schwartz) [reply mesonbuild/backend/ninjabackend.py:853]: https://github.com/mesonbuild/meson/blob/29c26d5b26397eaa3606d22d71309ecd1eb6b223/mesonbuild/compilers/compilers.py#L54-L61

It's not a routine, but it is a predefined lookup table.

Comment (dcbaker) [reply mesonbuild/backend/ninjabackend.py:853]: or you could do `s.split('.')[1] in compilers.lang_suffixes['d']`

Comment (rtbo) [reply mesonbuild/backend/ninjabackend.py]: yes OK. I used `s.split('.')[-1]` as there could be another point in the path

**Review by dcbaker — APPROVED** [review]

**Comment by tristan957** [comment]
@dcbaker is this good to merge? Maybe you were waiting for CI to pass and just forgot about it.

**Pull request merged**